File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/SortedArrayToBST.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.TreeNode;

public class SortedArrayToBST {
    public TreeNode sortedArrayToBST(int[] nums) {
        return buildTree(nums, 0, nums.length - 1);
    }

    private TreeNode buildTree(int[] nums, int leftIndex, int rightIndex) {
        if (leftIndex > rightIndex) return null;

        int midIndex = leftIndex + (rightIndex - leftIndex) / 2;
        TreeNode root = new TreeNode(nums[midIndex]);

        root.left = buildTree(nums, leftIndex, midIndex - 1);
        root.right = buildTree(nums, midIndex + 1, rightIndex);

        return root;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/sortedArrayToBST.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.TreeNode

fun sortedArrayToBST(nums: IntArray): TreeNode? {
    if (nums.isEmpty()) return null

    fun buildTree(startIndex: Int, endIndex: Int): TreeNode? {
        if (startIndex > endIndex) return null

        val midIndex = startIndex + (endIndex - startIndex) / 2

        return TreeNode(nums[midIndex]).apply {
            left = buildTree(startIndex, midIndex - 1)
            right = buildTree(midIndex + 1, endIndex)
        }
    }

    return buildTree(0, nums.lastIndex)
}
